File: easyAlgos/arrays/JS/maxProfit1.js
const maxProfit = (prices) => {
  let maxGain = 0; // time complexity: O(1)
  if (prices.length) {
    let minPrice = prices[0], // time complexity: O(1)
      count = 0, // time complexity: O(1)
      length = prices.length; // time complexity: O(1)
    // time complexity: O(n)
    while (count < length) {
      if (prices[count] < minPrice)
        minPrice = prices[count]; // time complexity: O(1)
      else if (prices[count] - minPrice > maxGain) {
        maxGain = prices[count] - minPrice; // time complexity: O(1)
      }
      ++count; // time complexity: O(1)
    }
  }
  return maxGain; // time complexity: O(1)
};

// time complexity: O(n)
// space complexity: O(1)

prices1 = [7, 1, 5, 3, 6, 4];
console.log(maxProfit(prices1)); // Output: 5

prices2 = [7, 6, 4, 3, 1];
console.log(maxProfit(prices2)); // Output: 0

prices3 = [2, 4, 1, 6];
console.log(maxProfit(prices3)); // Output 5

prices4 = [];
console.log(maxProfit(prices4)); // output 0

File: easyAlgos/arrays/Python/max_profit_1.py
def max_profit(prices: list[int]) -> int:
    max_gain = 0  # time complexity: O(1)
    if len(prices):  # time complexity: O(1)
        min_price = prices[0]  # time complexity: O(1)
        for _, stock_price in enumerate(prices, 1):  # time complexity: O(n)
            if stock_price < min_price:  # time complexity: O(1)
                min_price = stock_price  # time complexity: O(1)
            elif (stock_price - min_price) > max_gain:  # time complexity: O(1)
                max_gain = stock_price - min_price  # time complexity: O(1)
    return max_gain  # time complexity: O(1)


# time complexity: O(n)
# space complexity: O(1)

# Example Usage:
prices1 = [7, 1, 5, 3, 6, 4]
print(max_profit(prices1))  # Output: 5

prices2 = [7, 6, 4, 3, 1]
print(max_profit(prices2))  # Output: 0

prices3 = [2, 4, 1, 6]
print(max_profit(prices3))  # Output 5

prices4 = []
print(max_profit(prices4))  # output 0
